### PR TITLE
feat(warn): make the warning messages more explicit (close #7764)

### LINKED ISSUE #7764: Make the warning messages more explicit
### What problem does this feature solve?
When seeing warning message like `[Vue warn]: Invalid prop: type check failed for prop "value". Expected Number, got String.`, it would be great if Vue could display the actual values it encountered.

For instance, immediately seeing `2` and `'2'` would speed up development.

### What does the proposed API look like?
No API changes

### PR DESCRIPTION
**Pull request by yuriyalekseyev**
**What kind of change does this PR introduce?** (check at least one)

- [x] Feature

**Does this PR introduce a breaking change?** (check one)

- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Here is my solution for #7764.
I'm sure it's far from perfect, so I'll be happy to see any suggestion or advice 👍

### CODE REVIEW

File: src/core/util/props.js [hunk 1]
```diff
@@ -127,11 +127,41 @@ function assertProp (
...
+  function invalidTypeMessage () {
+    let message = `Invalid prop: type check failed for prop "${name}".` +
+      ` Expected ${expectedTypes.map(capitalize).join(', ')}`
+    const expectedValue = styleValue(value, expectedTypes[0])
+    const receivedValue = styleValue(value, toRawType(value))
+    const expectedType = expectedTypes[0].toLowerCase()
+    const receivedType = toRawType(value)
+    // check if we need to specify expected value
+    if (expectedTypes.length === 1 && isStringOrNumber(expectedType)) {
+      message += ` with value ${expectedValue}`
```
Comment (dennythecoder) [inline src/core/util/props.js]: This _is_ definitely an improvement on the message.  Couldn't boolean values be included as well?

**Comment by yuriyalekseyev** [comment]
@dennythecoder, thank you for the review! I have included boolean values as you suggested and done some refactoring.

File: src/core/util/props.js [hunk 2]
```diff
@@ -127,11 +127,48 @@ function assertProp (
       valid = assertedType.valid
     }
   }
+
+  function invalidTypeMessage () {
```
Comment (phanan) [inline src/core/util/props.js]: `invalidTypeMessage` sounds to me more like a property, not a function. How about `getInvalidTypeMessage()`?

Comment (yuriyalekseyev) [reply src/core/util/props.js]: @phanan, I agree, with a verb in front it sounds more correct, fixed it.

**Pull request merged**